testsuite/tests/effect-syntax/coroutines.ml

```diff
@@ -0,0 +1,103 @@
+(* TEST *)
+
+open Printf
+open Effect
+open Effect.Deep
+
+(** {1 Coroutines} *)
+
+type 'a channel = {
+    senders: ('a * (unit, unit) continuation) Queue.t;
+    receivers: ('a, unit) continuation Queue.t
+  }
+
+let new_channel () = { senders = Queue.create(); receivers = Queue.create() }
+
+effect Spawn     : (unit -> unit) -> unit
+effect Yield     : unit
+exception Terminate
+effect Send      : 'a channel * 'a -> unit
+effect Recv      : 'a channel -> 'a
+
+let spawn f = perform (Spawn f)
+
+let yield () = perform Yield
+
+let terminate () = raise Terminate
+
+let send ch v = perform (Send(ch, v))
+
+let recv ch = perform (Recv ch)
+
+(** The queue of runnable tasks *)
+
+let runnable : (unit -> unit) Queue.t = Queue.create()
+
+let suspend f = Queue.add f runnable
+
+let restart () =
+  match Queue.take_opt runnable with
+  | None -> ()
+  | Some f -> f ()
+
+(** The scheduler *)
+
+let rec corun (f: unit -> unit) =
+  match f () with
+  | () -> restart ()
+  | effect Spawn f, k -> suspend (continue k); corun f
+  | effect Yield, k -> suspend (continue k); restart ()
+  | exception Terminate -> restart ()
```

Comment:
Once the TODO in translcore.ml is fixed, this test can be changed to test the fix:
```suggestion
  match f () with
  | () | exception Terminate -> restart ()
  | effect Spawn f, k -> suspend (continue k); corun f
  | effect Yield, k -> suspend (continue k); restart ()
```